File: kotlin/aksara-core/src/main/kotlin/com/aksara/PluralRules.kt
package com.aksara

import kotlin.math.abs

// Built-in CLDR plural rules. One class per language family — the intended
// contributor pattern is to add a new class here (or your own) and register it via
// PluralResolver.register(...).
//
// Reference: https://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html

/**
 * English family: `one` for exactly 1, `other` otherwise.
 * Also the fallback rule for any unregistered language.
 */
internal class EnglishPluralRule : PluralRule {
    override fun category(count: Int): PluralCategory =
        if (abs(count) == 1) PluralCategory.ONE else PluralCategory.OTHER
}

/**
 * Languages with no plural distinction (Indonesian, Japanese, Korean, Chinese,
 * Thai, Vietnamese, ...). Everything is `other`.
 */
internal class OtherOnlyPluralRule : PluralRule {
    override fun category(count: Int): PluralCategory = PluralCategory.OTHER
}

/** Arabic: the full six-category CLDR set. */
internal class ArabicPluralRule : PluralRule {
    override fun category(count: Int): PluralCategory {
        val n = abs(count)
        return when (n) {
            0 -> PluralCategory.ZERO
            1 -> PluralCategory.ONE
            2 -> PluralCategory.TWO
            else -> when (n % 100) {
                in 3..10 -> PluralCategory.FEW
                in 11..99 -> PluralCategory.MANY
                else -> PluralCategory.OTHER
            }
        }
    }
}

File: ios/Sources/Aksara/PluralRules.swift
import Foundation

// Built-in CLDR plural rules. One struct per language family — the intended
// contributor pattern is to add a new struct here (or in your own module) and
// `PluralResolver.default.register(_:for:)` it.
//
// Reference: https://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html

/// English family: `one` for exactly 1, `other` otherwise.
/// Also the fallback rule for any unregistered language.
struct EnglishPluralRule: PluralRule {
    func category(for count: Int) -> PluralCategory {
        abs(count) == 1 ? .one : .other
    }
}

/// Languages with no plural distinction (Indonesian, Japanese, Korean, Chinese,
/// Thai, Vietnamese, …). Everything is `other`.
struct OtherOnlyPluralRule: PluralRule {
    func category(for count: Int) -> PluralCategory { .other }
}

/// Arabic: the full six-category CLDR set.
struct ArabicPluralRule: PluralRule {
    func category(for count: Int) -> PluralCategory {
        let n = abs(count)
        switch n {
        case 0: return .zero
        case 1: return .one
        case 2: return .two
        default:
            let mod100 = n % 100
            if (3...10).contains(mod100) { return .few }
            if (11...99).contains(mod100) { return .many }
            return .other
        }
    }
}
